libkbfs/block_tree.go

```diff
@@ -269,179 +275,235 @@ func (bt *blockTree) getNextDirtyBlockAtOffset(ctx context.Context,
+//   * blocks: a map from block pointer to a data-containing leaf node
+//     in the given range of offsets, if `getDirect` is true.
+//   * nextBlockOff is the offset of the block that follows the last
+//     block given in `pathsFromRoot`.  If `pathsFromRoot` contains
+//     the last block among the children, nextBlockOff is nil.
+func (bt *blockTree) getBlocksForOffsetRange(ctx context.Context,
+	ptr BlockPointer, pblock BlockWithPtrs, startOff, endOff Offset,
+	prefixOk bool, getDirect bool) (pathsFromRoot [][]parentBlockAndChildIndex,
+	blocks map[BlockPointer]Block, nextBlockOffset Offset,
+	err error) {
+	// Make a WaitGroup to keep track of whether there's still work to be done.
+	var wg sync.WaitGroup
+
+	// Make a workerpool to limit the number of concurrent goroutines.
+	wp := workerpool.New(maxBlockFetchWorkers)
+	defer wp.Stop() // TODO: this may be unnecessary
+
+	// Make a context to cancel all the jobs if something goes wrong
+	// TODO: determine if we can collapse this into the workerpool
+	groupCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	// Make a queue for results coming back from workers.
+	results := make(chan getBlocksForOffsetRangeResult)
+
+	// Enqueue the top-level task. Increment the task counter by one.
+	rootTask := getBlocksForOffsetRangeTask{
+		ptr:        ptr,
+		pblock:     pblock,
+		pathPrefix: nil,
+		startOff:   startOff,
+		endOff:     endOff,
+		prefixOk:   prefixOk,
+		getDirect:  getDirect,
 	}
-
+	wg.Add(1)
+	wp.Submit(func() {
+		bt.processGetBlocksTask(groupCtx, &wg, wp, rootTask, results)
+	})
+
+	// Once all the work is done, stop the WorkerPool and close `results` so
+	// that the loop below exits
+	go func() {
+		wg.Wait()
+		wp.Stop()
+		close(results)
+	}()
+
+	// Reduce all the results coming in over the results channel.
+	var minNextBlockOffset Offset
 	blocks = make(map[BlockPointer]Block)
-	var minNextBlockOffsetChild Offset
-outer:
-	for _, respCh := range respChans {
-		select {
-		case r := <-respCh:
-			pathsFromRoot = append(pathsFromRoot, r.pathsFromRoot...)
-			for ptr, block := range r.blocks {
-				blocks[ptr] = block
-			}
-			// We want to find the leftmost block offset that's to the
-			// right of the range, the one immediately following the
-			// end of the range.
-			if r.nextBlockOffset != nil &&
-				(minNextBlockOffsetChild == nil ||
-					r.nextBlockOffset.Less(minNextBlockOffsetChild)) {
-				minNextBlockOffsetChild = r.nextBlockOffset
-			}
-		default:
-			// There should always be a response ready in every
-			// channel, unless prefixOk is true.
-			if prefixOk {
-				break outer
-			} else {
-				panic("No response ready when !prefixOk")
+	pathsFromRoot = [][]parentBlockAndChildIndex{}
+	for res := range results {
+		if res.err != nil {
+			// If we are ok with just getting the prefix, don't treat a
+			// deadline exceeded error as fatal.
+			if prefixOk && res.err == context.DeadlineExceeded {
+				//TODO: determine that this is correct
```

Comment:
I don't think that `break` is the right thing here, since results could come out of order and there might still be some valid results waiting in the queue.  Probably `continue` would be better.

Even with `continue`, though, I think you're introducing a new bug.  The idea with `prefixOk` is that a file system read is allowed to return less data than was requested.  So if the call asked for 100 bytes, but the read only returned 20 bytes, that's ok and doesn't result in an error.  Since we're no longer processing results in order here, the returned set of blocks might contain (for example) one block at the start of the range, and one at the end.  Then other code (in `file_data.go`) would incorrectly fill zeroes in between those two blocks, when the correct behavior would have been just to return the data in the first block, ignoring the last one.

I'm not quite sure about the right way to fix it.  I think that, after you sort the paths below, you can check that the `childIndex` of the final element of each path increases by 1 until it reaches the number of pointers in the parent block, at which point the `childIndex` of the parent element must increase by 1.  And if you find a hole, and `prefixOk == true`, then just remove all the paths after the hole from the slice being returned.

Maybe there's a less code-intensive way of doing it, but I'm drawing a blank right now.